qa/L0_DALI_GPU_ensemble/client.py

```diff
@@ -51,57 +61,67 @@ def _get_input(batch, name):
     inp.set_data_from_numpy(batch)
     return inp
 
-  def run_inference(self, batches):
+  def test_infer(self, batches, it):
     assert(len(batches) == len(self.input_names))
     if (len(batches) > 1):
       for b in batches:
         assert b.shape[0] == batches[0].shape[0]
-    inputs = [self._get_input(batch, name) for batch, name in zip(batches, self.input_names)]
     outputs = [t_client.InferRequestedOutput(name) for name in self.output_names]
-    results = self.client.infer(model_name=self.model_name, inputs=inputs, outputs=outputs)
-    return [results.as_numpy(name) for name in self.output_names]
+    inputs = [self._get_input(batch, name) for batch, name in zip(batches, self.input_names)]
+    res = self.client.infer(model_name=self.model_name, inputs=inputs, outputs=outputs)
+    res_data = [res.as_numpy(name) for name in self.output_names]
+    return it, batches, res_data
 
   def run_tests(self, data, compare_to, n_infers=-1, eps=1e-7):
     generator = data if n_infers < 1 else islice(cycle(data), n_infers)
-    for it, batches in enumerate(generator):
-      results = self.run_inference(batches)
-      ref = compare_to(*batches)
-      assert(len(results) == len(ref))
-      for out_i, (out, ref_out) in enumerate(zip(results, ref)):
-        assert out.shape == ref_out.shape
-        if not np.allclose(out, ref_out, atol=eps):
-          print("Test failure in iteration", it)
-          print("Output", out_i)
-          print("Expected:\n", ref_out)
-          print("Actual:\n", out)
-          assert False
-      print('PASS iteration:', it)
+    for pack in grouper(self.concurrency, enumerate(generator)):
+      with ThreadPoolExecutor(max_workers=self.concurrency+4) as executor:
+        results_f = [executor.submit(self.test_infer, data, it) for it, data in pack]
+        for future in as_completed(results_f):
```

Comment:
Do this line needs to be in `with ThreadPoolExecutor` scope?

Reply:
Not an expert on that, but looking at the documentation, `as_completed` must be called under with ThreadPoolExecutor

Reply:
Sure